pkg/controllerhelpers/handlers.go

```diff
@@ -202,3 +203,36 @@ func (h *Handlers[T]) HandleDeleteWithDepth(depth int, obj any, enqueueFunc Enqu
 func (h *Handlers[T]) HandleDelete(obj any, enqueueFunc EnqueueFuncType) {
 	h.HandleDeleteWithDepth(2, obj, enqueueFunc)
 }
+
+// DeletedStronglyTypedFinalStateUnknown is like DeletedFinalStateUnknown, but
+// uses a types.NamespacedName as the key. This is useful for objects that are
+// namespaced, as it allows for more precise identification of the object.
+type DeletedStronglyTypedFinalStateUnknown struct {
+	Key types.NamespacedName
+	Obj interface{}
+}
+
+// DeletionHandlingTypedKeyFunc checks for
+// DeletedStronglyTypedFinalStateUnknown objects before calling
+// MetaNamespaceKeyFuncWithStrongType. This function is essentially the
+// same as the function DeletionHandlingMetaNamespaceKeyFunc
+// but it handles keys of the types.NamespacedName type.
+// This is required for queues that use strongly typed keys of the same type
+func DeletionHandlingTypedKeyFunc(obj interface{}) (types.NamespacedName, error) {
+	if d, ok := obj.(DeletedStronglyTypedFinalStateUnknown); ok {
+		return d.Key, nil
+	}
+	return MetaNamespaceKeyFuncWithStrongType(obj)
+}
+
+// MetaNamespaceKeyFuncWithStrongType is a KeyFunc that returns
+// types.NamespacedName. This function is similar to
+// MetaNamespaceKeyFunc, but it returns a structured key instead of a string.
+// It is useful for queues that use strongly typed keys of the same type.
+func MetaNamespaceKeyFuncWithStrongType(obj interface{}) (types.NamespacedName, error) {
```

Comment:
I don't think the name reflects what this function does, it's not any more "strongly typed" than a function returning a string is